Added Solidity -> ink! guide (#97)
paritytech/ink-docs: docs/intro/ink-vs-solidity.md

  - [`mapping declaration`](#mapping-declaration)
  - [`mapping usage`](#mapping-usage)
  - [`struct`](#struct)
  - [`assertions / requires`](#assertions--requires)
  - [`timestamp`](#timestamp)
  - [`contract caller`](#contract-caller)
  - [`contract's address`](#contracts-address)
  - [`bytes`](#bytes)
  - [`uint256`](#uint256)
  - [`payable`](#payable)
  - [`received deposit / payment`](#received-deposit--payment)
  - [`contract balance`](#contract-balance)
  - [`transfer tokens from contract`](#transfer-tokens-from-contract)
  - [`events & indexed`](#events--indexed)
  - [`errors and returning`](#errors-and-returning)
    - [`throw`](#throw)
    - [`assert`](#assert)
    - [`require and revert`](#require-and-revert)
  - [`nested mappings + custom / advanced structures`](#nested-mappings--custom--advanced-structures)
  - [`cross-contract calling`](#cross-contract-calling)
  - [`submit generic transaction / dynamic cross-contract calling`](#submit-generic-transaction--dynamic-cross-contract-calling)
- [Limitations of ink! v3](#limitations-of-ink-v3)
- [Troubleshooting Errors](#troubleshooting-errors)
- [unit testing (off-chain)](#unit-testing-off-chain)

> ## What is ink!
> 
> ink! is the smart contract language used in Substrate. It is built from Rust -- meaning that the great features of Rust are included in ink!. These features are ideal for a smart contract language. Furthermore, ink! is compiled to WebAssembly, which is high-performant, consistent, and very well researched. Furthermore, using Rust allows for inherent [safe math](https://ink.substrate.io/faq#overflow-safety) checks. ink! smart contracts use Rust's no_std. So, some common Rust implementations are not directly supported. However, ink! does have crates that reimplement common Rust code to work in the Substrate runtime.
> 

Reviewer: ```suggestion
## What is ink!

ink! is the smart contract language used in Substrate. It is built from Rust -- meaning that the great features of Rust are included in ink!. These features are ideal for a smart contract language. Furthermore, ink! is compiled to WebAssembly, which is high-performant, consistent, and very well researched. Furthermore, using Rust allows for inherent [safe math](https://ink.substrate.io/faq#overflow-safety) checks. ink! smart contracts use Rust's no_std. So, some common Rust implementations are not directly supported. However, ink! does have crates that reimplement common Rust code to work in the Substrate runtime.
```

Can be left out, as we're already in the ink! docs here.

> ## Setup
> 
> Follow the [prepare-your-first-contract](https://docs.substrate.io/tutorials/smart-contracts/prepare-your-first-contract/) tutorial.

Reviewer: ```suggestion
```

> ## Converting Solidity Contract to ink!

Reviewer: ```suggestion
## Converting a Solidity Contract to ink!
```

### 1. Generate New ink! Contract

> Run the following to generate ink! boilerplate code (flipper project)

Reviewer: ```suggestion
Run the following to generate ink! boilerplate code for ink!'s "Hello, World!" (the [`flipper`](https://github.com/paritytech/ink/tree/master/examples/flipper) contract))
```

#### `require and revert`

In Solidity, `require` is used for general (normal) errors -- such as errors that occur based on user input. `require` does have the option for an error string. `revert` is very similar to `require` except that `revert` will be called in `if ... else` chains. Both `require` and `revert` will revert the chain state. In ink!, `if ... { return Err(Error::SomeError) }` should be used for `require` or `revert`. When a `Result::Err` is returned in ink!, then all state is reverted.

> In general, `Result::Err` should be used when a _calling contract_ needs to know _why_ a function failed. Otherwise, `assert!` should be used as it has less overhead than a `Result`.

Author: Is the last sentence accurate? Based on the discussion [here](https://github.com/paritytech/ink/issues/1335).
